lib/os/fdtable.c

```diff
@@ -374,6 +374,47 @@ int zvfs_close(int fd)
 	return res;
 }
 
+int zvfs_dup(int fd, int *newfd)
+{
+	int ret;
+
+	if (_check_fd(fd) < 0) {
+		return -1;
+	}
+
+	(void)k_mutex_lock(&fdtable_lock, K_FOREVER);
+
+	if (newfd == NULL) {
+		/* dup() - just find lowest-numbered fd */
+		ret = _find_fd_entry();
+	} else {
+		/* dup2() - check if newfd is valid */
+		if (_check_fd(*newfd) < 0) {
+			ret = -1;
+		} else {
+			if (fdtable[fd].vtable->close) {
+				(void)fdtable[fd].vtable->close(fdtable[fd].obj);
+			}
+			ret = *newfd;
+		}
+	}
+
+	if (ret >= 0) {
+		/* Mark entry as used and initialize fields */
+		if (newfd == NULL) {
+			(void)z_fd_ref(ret);
+		}
+		fdtable[ret] = fdtable[fd];
+		fdtable[ret].offset = 0;
```

Comment:
shouldn't the new descriptor shares the same offset ?

Reply:
It's possible that after a dup operation, the new FD might share the same offset. Let me double-check this behaviour on Linux and I'll update if needed.

Reply:
Interesting - the POSIX [spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/dup.html) does not specify whether they share the same offset after `dup()`, but in Linux, that's the case (and it's also stated in the Linux [man page](https://man7.org/linux/man-pages/man2/dup.2.html)).

I think it's fine to follow the Linux convention here.

Reply:
Fixed!